File: ruby/isbn-verifier/isbn_verifier.rb
# frozen_string_literal: true

module IsbnVerifier
  CHECK = -1
  private_constant :CHECK
  CONVERT_CHECK = ->(check) { check == "X" ? 10 : check.to_i }
  private_constant :CONVERT_CHECK
  DASHES = "-"
  private_constant :DASHES
  INITIAL_SUM = 0
  private_constant :INITIAL_SUM
  ISBN_10 = /\A\d{9}(\d|X)\z/.freeze
  private_constant :ISBN_10
  MULTIPLE = 11
  private_constant :MULTIPLE
  WEIGHTS = (1..10).to_a.reverse.freeze
  private_constant :WEIGHTS
  WITHOUT_CHECK = (0..-2).freeze
  private_constant :WITHOUT_CHECK

  module_function

  def valid?(string)
    string = string.delete(DASHES)
    return false unless string.match?(ISBN_10)

    string
      .then(&method(:generate_isbn_integers))
      .zip(WEIGHTS)
      .reduce(INITIAL_SUM, &method(:isbn_formula))
      .modulo(MULTIPLE)
      .zero?
  end

  def generate_isbn_integers(string)
    check = CONVERT_CHECK.call(string[CHECK])

    string
      .chars[WITHOUT_CHECK]
      .map(&:to_i)
      .append(check)
  end
  private_class_method :generate_isbn_integers

  def isbn_formula(acc, (integer, weight))
    acc + integer * weight
  end
  private_class_method :isbn_formula
end

File: elixir/isbn-verifier/lib/isbn_verifier.ex
defmodule IsbnVerifier do
  @check -1
  @isbn_10 ~r/\A\d{9}[\dX]\z/
  @multiple 11
  @weights 10..1

  @doc """
    Checks if a string is a valid ISBN-10 identifier

    ## Examples

      iex> ISBNVerifier.isbn?("3-598-21507-X")
      true

      iex> ISBNVerifier.isbn?("3-598-2K507-0")
      false

  """
  @spec isbn?(String.t()) :: boolean
  def isbn?(isbn) do
    isbn = String.replace(isbn, "-", "")
    String.match?(isbn, @isbn_10) and valid?(isbn)
  end

  defp valid?(isbn) do
    isbn
    |> generate_isbn_integers()
    |> Enum.zip(@weights)
    |> Enum.reduce(0, &add_isbn_formula/2)
    |> rem(@multiple)
    |> Kernel.==(0)
  end

  defp generate_isbn_integers(isbn) do
    check =
      isbn
      |> String.at(@check)
      |> convert_check()

    isbn
    |> String.codepoints()
    |> Enum.drop(@check)
    |> Enum.map(&String.to_integer/1)
    |> Enum.concat(check)
  end

  defp convert_check("X"), do: [10]
  defp convert_check(check), do: [String.to_integer(check)]

  defp add_isbn_formula({integer, weight}, acc), do: acc + integer * weight
end
